File: C#/Other/ComplexNumber.cs
/*
 * Complex Number Implementation
 * Author: Андрій Будильников
 * 
 * This program provides a ComplexNumber class for working with complex numbers.
 */

using System;

public class ComplexNumber
{
    public double Real { get; set; }
    public double Imaginary { get; set; }

    // Constructor
    public ComplexNumber(double real = 0, double imaginary = 0)
    {
        Real = real;
        Imaginary = imaginary;
    }

    // String representation
    public override string ToString()
    {
        if (Imaginary >= 0)
        {
            return $"{Real} + {Imaginary}i";
        }
        else
        {
            return $"{Real} - {Math.Abs(Imaginary)}i";
        }
    }

    // Addition
    public static ComplexNumber operator +(ComplexNumber c1, ComplexNumber c2)
    {
        return new ComplexNumber(c1.Real + c2.Real, c1.Imaginary + c2.Imaginary);
    }

    public static ComplexNumber operator +(ComplexNumber c, double scalar)
    {
        return new ComplexNumber(c.Real + scalar, c.Imaginary);
    }

    public static ComplexNumber operator +(double scalar, ComplexNumber c)
    {
        return new ComplexNumber(c.Real + scalar, c.Imaginary);
    }

    // Subtraction
    public static ComplexNumber operator -(ComplexNumber c1, ComplexNumber c2)
    {
        return new ComplexNumber(c1.Real - c2.Real, c1.Imaginary - c2.Imaginary);
    }

    public static ComplexNumber operator -(ComplexNumber c, double scalar)
    {
        return new ComplexNumber(c.Real - scalar, c.Imaginary);
    }

    public static ComplexNumber operator -(double scalar, ComplexNumber c)
    {
        return new ComplexNumber(scalar - c.Real, -c.Imaginary);
    }

    // Multiplication
    public static ComplexNumber operator *(ComplexNumber c1, ComplexNumber c2)
    {
        // (a + bi) * (c + di) = (ac - bd) + (ad + bc)i
        double real = c1.Real * c2.Real - c1.Imaginary * c2.Imaginary;
        double imaginary = c1.Real * c2.Imaginary + c1.Imaginary * c2.Real;
        return new ComplexNumber(real, imaginary);
    }

    public static ComplexNumber operator *(ComplexNumber c, double scalar)
    {
        return new ComplexNumber(c.Real * scalar, c.Imaginary * scalar);
    }

    public static ComplexNumber operator *(double scalar, ComplexNumber c)
    {
        return new ComplexNumber(c.Real * scalar, c.Imaginary * scalar);
    }

    // Division
    public static ComplexNumber operator /(ComplexNumber c1, ComplexNumber c2)
    {
        // (a + bi) / (c + di) = [(a + bi)(c - di)] / (c² + d²)
        double denominator = c2.Real * c2.Real + c2.Imaginary * c2.Imaginary;
        if (denominator == 0)
        {
            throw new DivideByZeroException("Division by zero complex number");
        }

        double real = (c1.Real * c2.Real + c1.Imaginary * c2.Imaginary) / denominator;
        double imaginary = (c1.Imaginary * c2.Real - c1.Real * c2.Imaginary) / denominator;
        return new ComplexNumber(real, imaginary);
    }

    public static ComplexNumber operator /(ComplexNumber c, double scalar)
    {
        if (scalar == 0)
        {
            throw new DivideByZeroException("Division by zero");
        }
        return new ComplexNumber(c.Real / scalar, c.Imaginary / scalar);
    }

    public static ComplexNumber operator /(double scalar, ComplexNumber c)
    {
        double denominator = c.Real * c.Real + c.Imaginary * c.Imaginary;
        if (denominator == 0)
        {
            throw new DivideByZeroException("Division by zero complex number");
        }

        double real = (scalar * c.Real) / denominator;
        double imaginary = (-scalar * c.Imaginary) / denominator;
        return new ComplexNumber(real, imaginary);
    }

    // Equality
    public static bool operator ==(ComplexNumber c1, ComplexNumber c2)
    {
        if (ReferenceEquals(c1, null) && ReferenceEquals(c2, null)) return true;
        if (ReferenceEquals(c1, null) || ReferenceEquals(c2, null)) return false;
        return (c1.Real == c2.Real) && (c1.Imaginary == c2.Imaginary);
    }

    public static bool operator !=(ComplexNumber c1, ComplexNumber c2)
    {
        return !(c1 == c2);
    }

    public override bool Equals(object obj)
    {
        if (obj is ComplexNumber other)
        {
            return (Real == other.Real) && (Imaginary == other.Imaginary);
        }
        return false;
    }

    public override int GetHashCode()
    {
        return Real.GetHashCode() ^ Imaginary.GetHashCode();
    }

    // Conjugate
    public ComplexNumber Conjugate()
    {
        return new ComplexNumber(Real, -Imaginary);
    }

    // Magnitude
    public double Magnitude()
    {
        return Math.Sqrt(Real * Real + Imaginary * Imaginary);
    }

    // Phase
    public double Phase()
    {
        return Math.Atan2(Imaginary, Real);
    }

    // To polar form
    public Tuple<double, double> ToPolar()
    {
        return new Tuple<double, double>(Magnitude(), Phase());
    }

    // Create from polar coordinates
    public static ComplexNumber FromPolar(double magnitude, double phase)
    {
        double real = magnitude * Math.Cos(phase);
        double imaginary = magnitude * Math.Sin(phase);
        return new ComplexNumber(real, imaginary);
    }

    // Main method to demonstrate the complex number class
    public static void Main(string[] args)
    {
        // Create some complex numbers
        ComplexNumber z1 = new ComplexNumber(3, 4);   // 3 + 4i
        ComplexNumber z2 = new ComplexNumber(1, -2);  // 1 - 2i

        Console.WriteLine("Complex Numbers:");
        Console.WriteLine($"z1 = {z1}");
        Console.WriteLine($"z2 = {z2}");

        // Basic operations
        Console.WriteLine("\nBasic Operations:");
        Console.WriteLine($"z1 + z2 = {z1 + z2}");
        Console.WriteLine($"z1 - z2 = {z1 - z2}");
        Console.WriteLine($"z1 * z2 = {z1 * z2}");
        Console.WriteLine($"z1 / z2 = {z1 / z2}");

        // Properties
        Console.WriteLine("\nProperties:");
        Console.WriteLine($"Conjugate of z1 = {z1.Conjugate()}");
        Console.WriteLine($"Magnitude of z1 = {z1.Magnitude()}");
        Console.WriteLine($"Phase of z1 = {z1.Phase()} radians");

        // Polar form
        var polar = z1.ToPolar();
        Console.WriteLine($"Polar form of z1: magnitude = {polar.Item1}, phase = {polar.Item2} radians");

        // Create from polar
        ComplexNumber z3 = ComplexNumber.FromPolar(5, 0.927);  // Should be close to 3 + 4i
        Console.WriteLine($"From polar (5, 0.927): {z3}");
    }
}

File: C++/Other/ComplexNumber.cpp
/*
 * Complex Number Implementation
 * Author: Андрій Будильников
 * 
 * This program provides a ComplexNumber class for working with complex numbers.
 */

#include <iostream>
#include <cmath>
#include <stdexcept>
#include <string>
#include <sstream>

class ComplexNumber {
private:
    double real;
    double imag;

public:
    // Constructor
    ComplexNumber(double r = 0, double i = 0) : real(r), imag(i) {}

    // Getters
    double getReal() const { return real; }
    double getImag() const { return imag; }

    // Setters
    void setReal(double r) { real = r; }
    void setImag(double i) { imag = i; }

    // String representation
    std::string toString() const {
        std::stringstream ss;
        if (imag >= 0) {
            ss << real << " + " << imag << "i";
        } else {
            ss << real << " - " << -imag << "i";
        }
        return ss.str();
    }

    // Addition
    ComplexNumber operator+(const ComplexNumber& other) const {
        return ComplexNumber(real + other.real, imag + other.imag);
    }

    ComplexNumber operator+(double scalar) const {
        return ComplexNumber(real + scalar, imag);
    }

    friend ComplexNumber operator+(double scalar, const ComplexNumber& cn) {
        return ComplexNumber(cn.real + scalar, cn.imag);
    }

    // Subtraction
    ComplexNumber operator-(const ComplexNumber& other) const {
        return ComplexNumber(real - other.real, imag - other.imag);
    }

    ComplexNumber operator-(double scalar) const {
        return ComplexNumber(real - scalar, imag);
    }

    friend ComplexNumber operator-(double scalar, const ComplexNumber& cn) {
        return ComplexNumber(scalar - cn.real, -cn.imag);
    }

    // Multiplication
    ComplexNumber operator*(const ComplexNumber& other) const {
        // (a + bi) * (c + di) = (ac - bd) + (ad + bc)i
        double r = real * other.real - imag * other.imag;
        double i = real * other.imag + imag * other.real;
        return ComplexNumber(r, i);
    }

    ComplexNumber operator*(double scalar) const {
        return ComplexNumber(real * scalar, imag * scalar);
    }

    friend ComplexNumber operator*(double scalar, const ComplexNumber& cn) {
        return ComplexNumber(cn.real * scalar, cn.imag * scalar);
    }

    // Division
    ComplexNumber operator/(const ComplexNumber& other) const {
        // (a + bi) / (c + di) = [(a + bi)(c - di)] / (c² + d²)
        double denominator = other.real * other.real + other.imag * other.imag;
        if (denominator == 0) {
            throw std::runtime_error("Division by zero complex number");
        }

        double r = (real * other.real + imag * other.imag) / denominator;
        double i = (imag * other.real - real * other.imag) / denominator;
        return ComplexNumber(r, i);
    }

    ComplexNumber operator/(double scalar) const {
        if (scalar == 0) {
            throw std::runtime_error("Division by zero");
        }
        return ComplexNumber(real / scalar, imag / scalar);
    }

    friend ComplexNumber operator/(double scalar, const ComplexNumber& cn) {
        double denominator = cn.real * cn.real + cn.imag * cn.imag;
        if (denominator == 0) {
            throw std::runtime_error("Division by zero complex number");
        }

        double r = (scalar * cn.real) / denominator;
        double i = (-scalar * cn.imag) / denominator;
        return ComplexNumber(r, i);
    }

    // Equality
    bool operator==(const ComplexNumber& other) const {
        return (real == other.real) && (imag == other.imag);
    }

    bool operator==(double scalar) const {
        return (real == scalar) && (imag == 0);
    }

    friend bool operator==(double scalar, const ComplexNumber& cn) {
        return (cn.real == scalar) && (cn.imag == 0);
    }

    // Conjugate
    ComplexNumber conjugate() const {
        return ComplexNumber(real, -imag);
    }

    // Magnitude
    double magnitude() const {
        return std::sqrt(real * real + imag * imag);
    }

    // Phase
    double phase() const {
        return std::atan2(imag, real);
    }

    // To polar form
    std::pair<double, double> toPolar() const {
        return std::make_pair(magnitude(), phase());
    }

    // Create from polar coordinates
    static ComplexNumber fromPolar(double magnitude, double phase) {
        double r = magnitude * std::cos(phase);
        double i = magnitude * std::sin(phase);
        return ComplexNumber(r, i);
    }
};

// Overload << operator for easy printing
std::ostream& operator<<(std::ostream& os, const ComplexNumber& cn) {
    os << cn.toString();
    return os;
}

// Main function to demonstrate the complex number class
int main() {
    // Create some complex numbers
    ComplexNumber z1(3, 4);   // 3 + 4i
    ComplexNumber z2(1, -2);  // 1 - 2i

    std::cout << "Complex Numbers:" << std::endl;
    std::cout << "z1 = " << z1 << std::endl;
    std::cout << "z2 = " << z2 << std::endl;

    // Basic operations
    std::cout << "\nBasic Operations:" << std::endl;
    std::cout << "z1 + z2 = " << (z1 + z2) << std::endl;
    std::cout << "z1 - z2 = " << (z1 - z2) << std::endl;
    std::cout << "z1 * z2 = " << (z1 * z2) << std::endl;
    std::cout << "z1 / z2 = " << (z1 / z2) << std::endl;

    // Properties
    std::cout << "\nProperties:" << std::endl;
    std::cout << "Conjugate of z1 = " << z1.conjugate() << std::endl;
    std::cout << "Magnitude of z1 = " << z1.magnitude() << std::endl;
    std::cout << "Phase of z1 = " << z1.phase() << " radians" << std::endl;

    // Polar form
    auto polar = z1.toPolar();
    std::cout << "Polar form of z1: magnitude = " << polar.first 
              << ", phase = " << polar.second << " radians" << std::endl;

    // Create from polar
    ComplexNumber z3 = ComplexNumber::fromPolar(5, 0.927);  // Should be close to 3 + 4i
    std::cout << "From polar (5, 0.927): " << z3 << std::endl;

    return 0;
}